code has been improvised to generate fast response

iamananya opened the pull request:
@GuiltyMorishita the code have been improved for a fast response using hash tables.
![image](https://github.com/iamananya/ginco-task/assets/61614556/cd614000-d838-417b-a486-6353ed2b3dd4)

The implementation details are:
*  The `GenerateCharacterPool` function creates a hash table (probability map) to store the character probabilities. The character ID is used as the key in the hash table, and the probability is the associated value.
* The `FindCharacterByID` function iterates across the probability map and returns the character for each ID. The function directly retrieves the character using the ID from the hash table rather than scanning through the complete character slice for each character.
* The character pool is drawn at random using the `DrawCharacter` function

iamananya commented:
And for time=100000 the response time is:
![image](https://github.com/iamananya/ginco-task/assets/61614556/61d43a69-d2c5-4f34-bde2-e0c58612a039)

iamananya commented:
Also the results for `times=20` are:

 **R=16/20 which is 80% probability**  

 **SR=3/20 which is 15% probability**

 **SSR=1/20 which is 5% probability**
`The results are according to the probability mentioned in the code`. This proves the correctness of the code.

{
    "results": [
        {
            "characterID": "Character-46",
            "name": "Xiangling",
            "rarity": "R"
        },
        {
            "characterID": "Character-18",
            "name": "Venti",
            "rarity": "R"
        },
        {
            "characterID": "Character-28",
            "name": "Arataki Itto",
            "rarity": "R"
        },
        {
            "characterID": "Character-15",
            "name": "Mona",
            "rarity": "SR"
        },
        {
            "characterID": "Character-13",
            "name": "Kazuha",
            "rarity": "SR"
        },
        {
            "characterID": "Character-11",
            "name": "Hu Tao",
            "rarity": "R"
        },
        {
            "characterID": "Character-33",
            "name": "Zhongli",
            "rarity": "SR"
        },
        {
            "characterID": "Character-11",
            "name": "Hu Tao",
            "rarity": "R"
        },
        {
            "characterID": "Character-14",
            "name": "Kokomi",
            "rarity": "R"
        },
        {
            "characterID": "Character-27",
            "name": "Gorou",
            "rarity": "R"
        },
        {
            "characterID": "Character-46",
            "name": "Xiangling",
            "rarity": "R"
        },
        {
            "characterID": "Character-14",
            "name": "Kokomi",
            "rarity": "R"
        },
        {
            "characterID": "Character-46",
            "name": "Xiangling",
            "rarity": "R"
        },
        {
            "characterID": "Character-45",
            "name": "Barbara",
            "rarity": "R"
        },
        {
            "characterID": "Character-7",
            "name": "Water Nymph",
            "rarity": "R"
        },
        {
            "characterID": "Character-11",
            "name": "Hu Tao",
            "rarity": "R"
        },
        {
            "characterID": "Character-38",
            "name": "Raiden Shogun",
            "rarity": "SSR"
        },
        {
            "characterID": "Character-7",
            "name": "Water Nymph",
            "rarity": "R"
        },
        {
            "characterID": "Character-8",
            "name": "Ayaka",
            "rarity": "R"
        },
        {
            "characterID": "Character-46",
            "name": "Xiangling",
            "rarity": "R"
        }
    ]
}

GuiltyMorishita commented:
I confirm that it has been more efficient.
However, it appears that the user's character is not stored in the database, is that correct?

GuiltyMorishita commented:
@iamananya The proof is not enough.
The number of times is too small for 20 times, use the data of 10000 times.
You only looked at rarity, right? You also need to check if characters of the same rarity are draw at the same rate.

iamananya commented:
> @iamananya The proof is not enough. The number of times is too small for 20 times, use the data of 10000 times. You only looked at rarity, right? You also need to check if characters of the same rarity are draw at the same rate.

Yes I only checked the rarity. I will show you the proof of 10000 times data as well. 
I didn't understand the meaning of character of same rarity are drawn at same rate. Can you please explain @GuiltyMorishita

iamananya commented:
> I confirm that it has been more efficient. However, it appears that the user's character is not stored in the database, is that correct?

Yes they are not stored 🥲.

iamananya commented:
> I confirm that it has been more efficient. However, it appears that the user's character is not stored in the database, is that correct?

I have added in the latest commit @GuiltyMorishita .

GuiltyMorishita commented:
> I didn't understand the meaning of character of same rarity are drawn at same rate. Can you please explain

@iamananya 

I think it is the following logic.
- There are two S-ranked characters. (Character A, B).
- The draw rate of S-ranked characters is 10%.
- Therefore, the draw rate of character A is 5%.

Therefore, I want you to prove that each character is drawn according to the assumed probability.
And I want to make sure that it is not the case that only certain characters are draw more.

GuiltyMorishita commented on pkg/models/gacha.go:
```diff
@@ -136,33 +135,39 @@ func GetAllCharacters() ([]Character, error) {
+	return nil
+}
+
+func createBatch(tx *gorm.DB, batch []*UserCharacter) error {
```
This function name makes it difficult to understand what is being done from the outside.

iamananya replied:
Okay I will change the function name.

GuiltyMorishita commented on pkg/controllers/gacha-controller.go:
```diff
@@ -141,59 +143,46 @@ func HandleGachaDraw(w http.ResponseWriter, r *http.Request) {
-		i++
+	}
+	// Batch insert user characters into the database
+	if err := models.CreateUserCharacterBatch(userCharacters); err != nil {
```
Please name it in a way that makes it clear that you are not creating a batch, but rather creating characters with batch.

iamananya replied:
Okay I meant create user's character batch hence I named it that way.However I will change it if it doesn't make understandable.

GuiltyMorishita commented on pkg/controllers/gacha-controller.go:
```diff
@@ -141,59 +143,46 @@ func HandleGachaDraw(w http.ResponseWriter, r *http.Request) {
 		return
 	}
-	characterPool := generateCharacterPool(characters)
+	characterPool := GenerateCharacterPool(characters)
```
What is the reason for capitalization?

iamananya replied:
There is no reason, I just liked it that way :smile:

GuiltyMorishita replied:
@iamananya 
In Golang, case has meaning.
You need to look up the Golang specs.

iamananya replied:
@GuiltyMorishita Since this function is not exported, I will make it lowercase.

iamananya commented:
@GuiltyMorishita I have added all the changes suggested by you . Since I needed to prove the correctness. and I was getting some vague results in the previous implementation. Hence I have modified the code as well.

iamananya commented:
This PR fixes issue #9
